xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,477 @@
+    checkState(adsStream == null, "previous adsStream has not been cleared yet");
+    this.channel = checkNotNull(channel, "channel");
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
+        AggregatedDiscoveryServiceGrpc.newStub(channel);
+    adsStream = new AdsStream(stub);
+    adsStream.start();
+    stopwatch.reset().start();
+    adsStream.sendLdsRequest(targetName, "");
+  }
+
+  @VisibleForTesting
+  void shutdownDiscoveryRpc() {
+    if (adsStream != null) {
+      adsStream.close(Status.CANCELLED.withDescription("shutdown").asException());
+    }
+    if (rpcRetryTimer != null) {
+      rpcRetryTimer.cancel();
+    }
+  }
+
+  /**
+   * Handles LDS response to find the HttpConnectionManager message for the requested resource name.
+   * Sends an ACK request to the management server if the HttpConnectionManager for the requested
+   * resource name is found, which may either contain the RouteConfiguration directly in-line or
+   * contain the configuration to send RDS requests for dynamic resolution. Otherwise, sends back
+   * a NACK request. Proceed with the resolved RouteConfiguration (if exists) to find the
+   * VirtualHost configuration for the "xds:" URI (with the port, if any, stripped off).
+   */
+  private void handleLdsResponse(DiscoveryResponse ldsResponse) {
+    logger.log(Level.FINE, "Received a LDS response: {0}", ldsResponse);
+    List<Any> respResources = ldsResponse.getResourcesList();
+    HttpConnectionManager connManager = null;
+    try {
+      for (com.google.protobuf.Any res : respResources) {
+        Listener listener = res.unpack(Listener.class);
+        if (listener.getName().equals(targetName)) {
+          HttpConnectionManager httpConnManager =
+              listener.getApiListener().getApiListener().unpack(HttpConnectionManager.class);
+          if (httpConnManager.hasRouteConfig() || httpConnManager.hasRds()) {
+            connManager = httpConnManager;
+          }
+          // Ignore remaining listeners once the requested one is found.
+          break;
+        }
+      }
+    } catch (InvalidProtocolBufferException e) {
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Invalid LDS response"));
+    }
+    if (connManager != null) {
+      // Accept this listener update.
+      adsStream.ldsVersion = ldsResponse.getVersionInfo();
+    }
+    // Send ACK/NACK request.
+    adsStream.sendLdsRequest(targetName, ldsResponse.getNonce());
+    if (connManager != null) {
+      if (connManager.hasRouteConfig()) {
+        processRouteConfig(connManager.getRouteConfig());
+      } else if (connManager.hasRds()) {
+        String rcName = connManager.getRds().getRouteConfigName();
+        adsStream.sendRdsRequest(rcName, ldsResponse.getNonce());
+      } else {
+        // Impossible to be here.
+        throw new AssertionError("Severe bug: accepted listener update contains invalid config");
```

Comment:
I think we should send NACK instead of panic.

Reply:
This is not a panic. Code can never ever reach here (it's a placeholder for the `else` branch). If reaching here, the code is very wrong.

Reply:
Can "wrong" input cause this as well? For example, certain combination of values in `connManager`? Also shouldn't we delay sending ACK/NACK (line 188) until after these steps are done?

Reply:
How can it never ever reach here. If the server of some version of TD/envoy management server does not support this feature, and it gives a default instance of HttpConnectionManager. Then it may reach here?

Reply:
Improved error handling.